src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2616 @@
+            object_len++;
+        }
+        if (offset + object_len > len) {
+            SCLogInfo("Insufficient data to decode object %d:%d", header->group,
+                header->variation);
+            goto not_enough_data;
+        }
+
+        offset += object_len;
+    next:
+
+        object->data = SCCalloc(1, offset);
+        if (object->data != NULL) {
+            memcpy(object->data, buf, offset);
+            object->len = offset;
+        }
+
+        buf += offset;
+        len -= offset;
+
+        continue;
+
+        /* Can't continue on an unknown group or variation as we don't
+         * know how far to advance the buffer. */
+
+    unknown:
+        SCLogInfo("Unknown variation 0x%02x (%d) for group 0x%02x (%d)",
+            header->variation, header->variation, header->group, header->group);
+        goto done;
+    }
+
+    /* All objects were decoded. */
+    retval = 1;
+
+not_enough_data:
+done:
+    return retval;
+}
+
+static int DNP3HandleRequestTransportLayer(DNP3State *dnp3, uint8_t *input,
+    uint32_t input_len)
+{
+    SCEnter();
+    DNP3LinkHeader *header;
+    DNP3Session *session;
+    DNP3TransportHeader th;
+    DNP3Transaction *tx;
+    DNP3ApplicationHeader *ah;
+    uint32_t offset = 0;
+    uint32_t th_offset = 0;
+
+    header = (DNP3LinkHeader *)input;
+    offset += sizeof(DNP3LinkHeader);
+
+    th = input[offset];
+    th_offset = offset++;
+
+    session = DNP3SessionGet(dnp3, header);
+    if (unlikely(session == NULL)) {
+        goto error;
+    }
+
+    if (!DNP3_TRANSPORT_FIR(th)) {
+        /* Need to look further into multi-segment fragments from the master. */
+        SCLogInfo("Unexpected multi-frame fragment from master.");
+        goto done;
+    }
+
+    ah = (DNP3ApplicationHeader *)(input + offset);
+    offset += sizeof(DNP3ApplicationHeader);
+
+    tx = DNP3TxAlloc(dnp3);
+    if (unlikely(tx == NULL)) {
+        goto error;
+    }
+    tx->session = session;
+    session->last_tx = tx;
+    tx->app_seqno = DNP3_APP_SEQ(ah->control);
+    tx->app_function_code = ah->function_code;
+    tx->request_ll_control = header->control;
+    tx->request_th = th;
+    tx->request_al_control = ah->control;
+    tx->request_al_fc = ah->function_code;
+
+    /* Some function codes do not expect a reply. */
+    switch (ah->function_code) {
+        case DNP3_APP_FC_CONFIRM:
+        case DNP3_APP_FC_DIR_OPERATE_NR:
+        case DNP3_APP_FC_FREEZE_NR:
+        case DNP3_APP_FC_FREEZE_CLEAR_NR:
+        case DNP3_APP_FC_FREEZE_AT_TIME_NR:
+        case DNP3_APP_FC_AUTH_REQ_NR:
+            tx->response_done = 1;
+        default:
+            break;
+    }
+
+    /* Reassemble the request buffer. */
+    tx->request_buffer = DNP3ReassembleApplicationLayer(
+        input + th_offset, // + sizeof(DNP3TransportHeader),
+            input_len - th_offset, // - sizeof(DNP3TransportHeader),
+            &tx->request_buffer, &tx->request_buffer_len);
+    tx->request_done = 1;
+
```

Comment:
In next PR, the return code of DNP3ReassembleApplicationLayer will be checked, and a malformed event will be raised as needed.